cmd/exec.go

```diff
@@ -306,3 +347,94 @@ func isStaleFile(file os.FileInfo) bool {
 
 	return time.Now().Sub(file.ModTime()) > timeLimit
 }
+
+func gzipLogData(logData string) *bytes.Buffer {
+	var b bytes.Buffer
+	if len(logData) < 1 {
+		return &b
+	}
+
+	gz := gzip.NewWriter(&b)
+	if _, err := gz.Write([]byte(logData)); err != nil {
+		log("error writing gzip")
+		return nil
+	}
+	if err := gz.Close(); err != nil {
+		log("error closing gzip")
+		return nil
+	}
+	return &b
+}
+
+func getPresignedUrl(postBody []byte) ([]byte, error) {
```

Comment:
would be nice to move this into lib/cronitor to keep this exec command focused on its main job. That module currently uses our v3 api end points but I will eventually update them to be consistent with this.

Reply:
A few thoughts on this:

* As you said, currently `lib/cronitor` is built very differently than that bit is, including API versions (presign doesn't even have a version), etc.
* Personally I think the presign API call actually acts quite a bit differently than `lib/cronitor` and the rest of the API. It is _technically_ an API call, but only because we need to keep the AWS keys secret in order to presign. I think there's another valid argument that it actually belongs inside `exec`, because its sole purpose is to presign URLs for log uploading, which is very much an `exec`-related task.

Given the urgency I understand around shipping this, I suggest we leave this as is for now, and then if you disagree with what I wrote above and really want me to move it to `lib/cronitor`, I can refactor after.

Reply:
i think 100% it took you longer to write this reply than to use your IDE refactoring tool to move the method. 

There is a little organization to this code right now and that is, all of the code that hits cronitor is in lib/cronitor or sendPing -- we should follow that. Eventually i will move the whole module to use the new API which removes versions from the URLs so they look 100% identical to your presign URL (versioning done via header). 

If you are pressed for time I can push a change to your branch let me know.

Reply:
Ok I just pushed this, i was hung up at the end with a problem but the I realize it failed still when I stashed my changes.

Reply:
Not a time thing--philosophically I genuinely think it should sit in `exec`. But I'm happy to move it.

Regarding shipping now vs later, I misunderstood what you were saying--I thought you were suggesting refactoring the methods to fit as methods of the `CronitorApi` struct that appears in `lib/cronitor`--which actually if we did it would be clean. cronitor-kubernetes has a very similar structure and it looks nice.  And actually looking at what I did [there](https://github.com/cronitorio/cronitor-kubernetes/blob/main/pkg/api/logs.go#L48), I actually did attach `shipLogData` to the main struct, so I'm probably disproving my own earlier point given how clean that looks, and maybe it should be refactored.

That said if you literally just meant copy/pasting it to `lib/cronitor`, obviously that will take like 5 seconds. I can do that now.

Reply:
Here's the error currently:

```
mac: ~/0x0/cli $ ./cronitor-cli -v exec shane-testy2 'true && echo "test"'
Reading config from /etc/cronitor/cronitor.json
Running subcommand: true && echo "test"
Sending ping https://cronitor.link/ping/85dfb5365b3446c280fa6448cda91a94/shane-testy2?state=run&try=1&stamp=1651207111.322&msg=true+%26%26+echo+%22test%22&host=Shanes-MacBook-Pro.local&series=1651207111.322
test
Sending ping https://cronitor.link/ping/85dfb5365b3446c280fa6448cda91a94/shane-testy2?state=complete&try=1&stamp=1651207111.351&msg=test%0A&host=Shanes-MacBook-Pro.local&duration=0.029&series=1651207111.322&status_code=0&metric=length%3A5
error generating presign url for log uploading: error response code 400 returned
```

Reply:
the pings are correct, has the correct length, etc, but the log upload is not happy
https://cronitor.io/app/monitors/shane-testy2?env=production&time=7d